tools/congestion-model/src/strategy/smooth_traffic_light.rs

```diff
@@ -0,0 +1,352 @@
+        }
+    }
+
+    // Checks if the transaction receiver is in a congested shard. If so the
+    // transaction should be rejected.
+    fn get_filter_stop(&mut self, ctx: &mut ChunkExecutionContext<'_>, tx: TransactionId) -> bool {
+        let receiver = ctx.tx_receiver(tx);
+
+        let CongestedShardsInfo { congestion_level, .. } = self.get_info(ctx, &receiver);
+        congestion_level > self.reject_tx_congestion_limit
+    }
+
+    // Step 4: Execute receipts in the order of local, delayed, incoming
+    // Step 5: Remaining local or incoming receipts are added to the end of the
+    // delayed receipts queue
+    //
+    // In the model there is no distinction between local, delayed and incoming.
+    // All of those are stored in the incoming queue so we just process that.
+    //
+    // Always process as many receipts as allowed by the GAS_LIMIT.
+    //
+    // The outgoing receipts are processed as in `process_outgoing_receipts`.
+    fn process_incoming_receipts(&mut self, ctx: &mut ChunkExecutionContext<'_>) {
+        while ctx.gas_burnt() < GAS_LIMIT {
+            let Some(receipt) = ctx.incoming_receipts().pop_front() else {
+                break;
+            };
+
+            let outgoing = ctx.execute_receipt(receipt);
+            for receipt in outgoing {
+                self.forward_or_buffer(ctx, receipt);
+            }
+        }
+    }
+
+    // Step 6: Compute own congestion information for the next block
+    fn update_block_info(&mut self, ctx: &mut ChunkExecutionContext<'_>) {
+        let incoming_gas_congestion = self.incoming_gas_congestion(ctx);
+        let outgoing_gas_congestion = self.outgoing_gas_congestion(ctx);
+        let memory_congestion = self.memory_congestion(ctx);
+
+        // Initial traffic light:
+        // Usually, signal other shards to slow down based on our incoming gas congestion only.
+        // All other limits (outgoing gas, memory) are ignored until they hit a red light.
+        // The goal is to never hit red, unless we have to prevent the unbounded queues vs deadlocks tradeoff.
+        // If we hit red, it slows down incoming traffic massively, which should bring us out of red soon.
+        // let red =
+        //     incoming_gas_congestion.max(outgoing_gas_congestion).max(memory_congestion) >= 1.0;
+        // let info = if red {
+        //     CongestedShardsInfo {
+        //         congestion_level: 1.0,
+        //         allowed_shard: Some(self.round_robin_shard(ctx.block_height() as usize)),
+        //     }
+        // } else {
+        //     CongestedShardsInfo { congestion_level: incoming_gas_congestion, allowed_shard: None }
+        // };
+
+        // Simplified and smoothed. Requires a larger max memory limit but will
+        // smoothly reduce based on size, which can even lead to smaller peak
+        // buffer size.
+        let max_congestion =
+            incoming_gas_congestion.max(outgoing_gas_congestion).max(memory_congestion);
```

Comment:
cool syntax, I've been looking for a concise 3max and I arrived at vec![a, b, c].iter().max() but this is neat too!

Reply:
I think you don't even need `vec!`, thus you can avoid the allocation and it looks a bit shorter.
But I guess I'd still prefer avoiding the iterator in most cases.
Furthermore, your method doesn't work for `f32` and `f64` because they don't implement `Ord`... While the [`max`](https://doc.rust-lang.org/std/primitive.f32.html#method.max) methods on the floating point number types work. They simply drop NaNs if they occur, which is usually good enough.